File: data_baseline/openrouter__qwen__qwen3-coder-plus/python/rep4/todo_server.py
#!/usr/bin/env python3
"""
REST API server for managing personal todo items with cookie-based authentication
"""
import json
import re
import uuid
from datetime import datetime, timezone
from http.cookies import SimpleCookie
from typing import Dict, List, Optional
from urllib.parse import parse_qs, urlparse

from http.server import HTTPServer, BaseHTTPRequestHandler

# In-memory storage
users = {}
todos = {}
sessions = {}
next_user_id = 1
next_todo_id = 1

class User:
    def __init__(self, user_id: int, username: str, password: str):
        self.id = user_id
        self.username = username
        self.password = password # In production, this should be hashed
        self.todo_ids = []

class Todo:
    def __init__(self, todo_id: int, user_id: int, title: str, description: str = ""):
        self.id = todo_id
        self.user_id = user_id  # Reference to user who owns the todo
        self.title = title
        self.description = description
        self.completed = False
        self.created_at = datetime.now(timezone.utc).strftime('%Y-%m-%dT%H:%M:%SZ')
        self.updated_at = datetime.now(timezone.utc).strftime('%Y-%m-%dT%H:%M:%SZ')

def validate_username(username: str) -> bool:
    """Check if username matches the pattern ^[a-zA-Z0-9_]+$
    and has length between 3-50 characters."""
    return re.match(r'^[a-zA-Z0-9_]{3,50}$', username) is not None

def get_current_timestamp():
    """Get current UTC timestamp in the required format."""
    return datetime.now(timezone.utc).strftime('%Y-%m-%dT%H:%M:%SZ')

def authenticate_request(headers) -> Optional[int]:
    """Extract session ID from cookies and return associated user ID,
    or None if authentication fails."""
    auth_header = headers.get('Cookie')
    if not auth_header:
        return None
    
    cookies = SimpleCookie()
    cookies.load(auth_header)
    
    session_cookie = cookies.get('session_id')
    if not session_cookie:
        return None
    
    session_id = session_cookie.value
    return sessions.get(session_id)

class TodoRequestHandler(BaseHTTPRequestHandler):
    
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        
    def _set_headers(self, status_code=200, content_type='application/json'):
        """Set response headers."""
        self.send_response(status_code)
        if content_type:
            self.send_header('Content-Type', content_type)
        self.end_headers()
    
    def _send_json_response(self, data, status_code=200):
        """Send JSON response."""
        self._set_headers(status_code)
        if data is not None:
            self.wfile.write(json.dumps(data).encode())
    
    def _send_no_content(self, status_code=204):
        """Send 204 No Content response."""
        self.send_response(status_code)
        self.end_headers()
    
    def _get_body(self):
        """Get request body as JSON."""
        content_length = int(self.headers.get('Content-Length', 0))
        body = self.rfile.read(content_length)
        return json.loads(body.decode()) if body else {}
    
    def _require_auth(self):
        """Check if request is authenticated."""
        user_id = authenticate_request(self.headers)
        if not user_id:
            return False, None
        return True, user_id
    
    def _check_todo_access(self, todo_id: int, user_id: int) -> bool:
        """Check if a user has access to a particular todo."""
        todo = todos.get(todo_id)
        if not todo:
            return False
        return todo.user_id == user_id
    
    def do_GET(self):
        """Handle GET requests."""
        parsed_path = urlparse(self.path)
        path_parts = parsed_path.path.strip('/').split('/')
        
        if path_parts[0] == 'me':
            # Check authentication
            auth, user_id = self._require_auth()
            if not auth:
                self._send_json_response({'error': 'Authentication required'}, 401)
                return
            
            # Get user by ID
            user = users.get(user_id)
            if not user:
                self._send_json_response({'error': 'User not found'}, 500)  # Should never happen
                return
            
            result = {
                'id': user.id,
                'username': user.username
            }
            self._send_json_response(result)
            
        elif len(path_parts) == 2 and path_parts[0] == 'todos':
            # Get specific todo
            try:
                todo_id = int(path_parts[1])
            except ValueError:
                self._send_json_response({'error': 'Invalid todo ID'}, 400)
                return
            
            # Check authentication
            auth, user_id = self._require_auth()
            if not auth:
                self._send_json_response({'error': 'Authentication required'}, 401)
                return
                
            # Check if user can access this todo
            if not self._check_todo_access(todo_id, user_id):
                self._send_json_response({'error': 'Todo not found'}, 404)
                return
            
            # Return todo
            todo = todos.get(todo_id)
            if not todo:
                self._send_json_response({'error': 'Todo not found'}, 404)
                return
            self._send_json_response(todo.__dict__)
            
        elif path_parts[0] == 'todos':
            # Check authentication
            auth, user_id = self._require_auth()
            if not auth:
                self._send_json_response({'error': 'Authentication required'}, 401)
                return
            
            # List all todos for this user
            user = users.get(user_id)
            if not user:
                self._send_json_response([], 200)
                return
            user_todos = [todo.__dict__ for todo_id in user.todo_ids 
                          for todo in [todos.get(todo_id)] 
                          if todo is not None]
            self._send_json_response(user_todos)
            
        else:
            self._send_json_response({'error': 'Endpoint not found'}, 404)
    
    def do_POST(self):
        """Handle POST requests."""
        parsed_path = urlparse(self.path)
        path_parts = parsed_path.path.strip('/').split('/')
        
        if path_parts[0] == 'register':
            # Handle registration
            try:
                request_data = self._get_body()
            except json.JSONDecodeError:
                self._send_json_response({'error': 'Invalid JSON in request'}, 400)
                return
            
            username = request_data.get('username')
            password = request_data.get('password')
            
            if not username:
                self._send_json_response({'error': 'Invalid username'}, 400)
                return
                
            if not validate_username(username):
                self._send_json_response({'error': 'Invalid username'}, 400)
                return
                
            if not password or len(password) < 8:
                self._send_json_response({'error': 'Password too short'}, 400)
                return
            
            # Check if username already exists
            for user in users.values():
                if user.username == username:
                    self._send_json_response({'error': 'Username already exists'}, 409)
                    return
            
            # Create new user (auto-increment ID)
            global next_user_id
            user_id = next_user_id
            next_user_id += 1
            new_user = User(user_id, username, password)
            users[user_id] = new_user
            
            result = {
                'id': new_user.id,
                'username': new_user.username
            }
            self._send_json_response(result, 201)
        
        elif path_parts[0] == 'login':
            # Handle login
            try:
                request_data = self._get_body()
            except json.JSONDecodeError:
                self._send_json_response({'error': 'Invalid JSON in request'}, 400)
                return
            
            username = request_data.get('username')
            password = request_data.get('password')
            
            # Find user by username
            target_user = None
            for user in users.values():
                if user.username == username:
                    target_user = user
                    break
                    
            if not target_user or target_user.password != password:
                self._send_json_response({'error': 'Invalid credentials'}, 401)
                return
            
            # Generate session
            session_id = str(uuid.uuid4())
            sessions[session_id] = target_user.id
            
            # Send response with SET_COOKIE header
            result = {
                'id': target_user.id,
                'username': target_user.username
            }
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Set-Cookie', f'session_id={session_id}; Path=/; HttpOnly')
            self.end_headers()
            self.wfile.write(json.dumps(result).encode())
        
        elif path_parts[0] == 'logout':
            # Check authentication
            auth, user_id = self._require_auth()
            if not auth:
                self._send_json_response({'error': 'Authentication required'}, 401)
                return
            
            # Remove session for current user (based on cookie)
            auth_header = self.headers.get('Cookie')
            if auth_header:
                cookies = SimpleCookie()
                cookies.load(auth_header)
                
                session_cookie = cookies.get('session_id')
                if session_cookie:
                    session_id = session_cookie.value
                    if session_id in sessions:
                        del sessions[session_id]
                        
            self._send_json_response({})
        
        elif path_parts[0] == 'todos':
            # Check authentication
            auth, user_id = self._require_auth()
            if not auth:
                self._send_json_response({'error': 'Authentication required'}, 401)
                return
            
            # Create new todo
            try:
                request_data = self._get_body()
            except json.JSONDecodeError:
                self._send_json_response({'error': 'Invalid JSON in request'}, 400)
                return
                
            title = request_data.get('title')
            description = request_data.get('description') or ""
            
            if not title:
                self._send_json_response({'error': 'Title is required'}, 400)
                return
                
            # Create new todo
            global next_todo_id
            todo_id = next_todo_id
            next_todo_id += 1
            new_todo = Todo(todo_id, user_id, title, description)
            todos[todo_id] = new_todo
            users[user_id].todo_ids.append(todo_id)
            
            self._send_json_response(new_todo.__dict__, 201)
        
        else:
            self._send_json_response({'error': 'Endpoint not found'}, 404)
    
    def do_PUT(self):
        """Handle PUT requests."""
        parsed_path = urlparse(self.path)
        path_parts = parsed_path.path.strip('/').split('/')
        
        if path_parts[0] == 'password':
            # Update password
            auth, user_id = self._require_auth()
            if not auth:
                self._send_json_response({'error': 'Authentication required'}, 401)
                return
            
            try:
                request_data = self._get_body()
            except json.JSONDecodeError:
                self._send_json_response({'error': 'Invalid JSON in request'}, 400)
                return
            
            old_password = request_data.get('old_password')
            new_password = request_data.get('new_password')
            
            # Validate current password
            user = users.get(user_id)
            if user.password != old_password:
                self._send_json_response({'error': 'Invalid credentials'}, 401)
                return
                
            # Validate new password length
            if not new_password or len(new_password) < 8:
                self._send_json_response({'error': 'Password too short'}, 400)
                return
                
            # Update password
            user.password = new_password
            self._send_json_response({})
        
        elif len(path_parts) == 2 and path_parts[0] == 'todos':
            # Check authentication
            auth, user_id = self._require_auth()
            if not auth:
                self._send_json_response({'error': 'Authentication required'}, 401)
                return
                
            # Update specific todo
            try:
                todo_id = int(path_parts[1])
            except ValueError:
                self._send_json_response({'error': 'Invalid todo ID'}, 400)
                return
            
            # Check if user can access this todo
            if not self._check_todo_access(todo_id, user_id):
                self._send_json_response({'error': 'Todo not found'}, 404)
                return
            
            try:
                request_data = self._get_body()
            except json.JSONDecodeError:
                self._send_json_response({'error': 'Invalid JSON in request'}, 400)
                return
            
            # Get the todo object
            todo = todos.get(todo_id)
            if not todo:
                self._send_json_response({'error': 'Todo not found'}, 404)
                return
            
            # Update fields that exist in the request
            # Handle updating title (if provided)
            if 'title' in request_data:
                title = request_data['title']
                if not title:
                    self._send_json_response({'error': 'Title is required'}, 400)
                    return
                todo.title = title
            
            # Handle updating description (if provided)
            if 'description' in request_data:
                todo.description = request_data['description']
            
            # Handle setting completed status (if provided)
            if 'completed' in request_data:
                completed = request_data['completed']
                if not isinstance(completed, bool):
                    self._send_json_response({'error': 'Completed field must be boolean'}, 400)
                    return
                todo.completed = completed
            
            # Update the timestamp
            todo.updated_at = get_current_timestamp()
            self._send_json_response(todo.__dict__)
        
        else:
            self._send_json_response({'error': 'Endpoint not found'}, 404)
    
    def do_DELETE(self):
        """Handle DELETE requests."""
        parsed_path = urlparse(self.path)
        path_parts = parsed_path.path.strip('/').split('/')
        
        if len(path_parts) == 2 and path_parts[0] == 'todos':
            # Check authentication
            auth, user_id = self._require_auth()
            if not auth:
                self._send_json_response({'error': 'Authentication required'}, 401)
                return
            
            # Delete specific todo
            try:
                todo_id = int(path_parts[1])
            except ValueError:
                self._send_json_response({'error': 'Invalid todo ID'}, 400)
                return
            
            # Check if user can access this todo
            if not self._check_todo_access(todo_id, user_id):
                self._send_json_response({'error': 'Todo not found'}, 404)
                return
                
            # Remove from user's todo list and delete the todo
            user = users.get(user_id)
            if user and todo_id in user.todo_ids:
                user.todo_ids.remove(todo_id)
            if todo_id in todos:
                del todos[todo_id]
            
            self._send_no_content(204)
        
        else:
            self._send_json_response({'error': 'Endpoint not found'}, 404)


def run_server(port):
    """Run the HTTP server on the specified port."""
    server_address = ('0.0.0.0', port)
    httpd = HTTPServer(server_address, TodoRequestHandler)
    print(f'Starting server on {server_address[0]}:{server_address[1]}...')
    httpd.serve_forever()


if __name__ == '__main__':
    import argparse
    
    parser = argparse.ArgumentParser(description='Todo REST API Server')
    parser.add_argument('--port', type=int, required=True, help='Port to listen on')
    
    args = parser.parse_args()
    run_server(args.port)

File: data_baseline/openrouter__qwen__qwen3-coder-plus/scala/rep4/src/main/scala/TodoServer.scala
import zhttp.core._
import zhttp.http._
import zhttp.service.{Client, Server, ChannelFactory, EventLoopGroup, SSLConfig}
import zio._
import zio.json._
import java.util.UUID
import java.time.format.DateTimeFormatter
import java.time.LocalDateTime
import java.time.ZoneOffset

// Data models
case class User(id: Int, username: String, passwordHash: String)
object User {
  implicit val codec: JsonCodec[User] = DeriveJsonCodec.gen[User]
}

case class CreateUser(username: String)
object CreateUser {
  implicit val codec: JsonCodec[CreateUser] = DeriveJsonCodec.gen[CreateUser]
}

case class LoginRequest(username: String, password: String)
object LoginRequest {
  implicit val codec: JsonCodec[LoginRequest] = DeriveJsonCodec.gen[LoginRequest]
}

case class ChangePassword(old_password: String, new_password: String)
object ChangePassword {
  implicit val codec: JsonCodec[ChangePassword] = DeriveJsonCodec.gen[ChangePassword]
}

case class CreateTodo(title: String, description: String)
object CreateTodo {
  implicit val codec: JsonCodec[CreateTodo] = DeriveJsonCodec.gen[CreateTodo]
}

case class UpdateTodo(title: Option[String], description: Option[String], completed: Option[Boolean])
object UpdateTodo {
  implicit val codec: JsonCodec[UpdateTodo] = DeriveJsonCodec.gen[UpdateTodo]
}

case class Todo(id: Int, title: String, description: String, completed: Boolean, created_at: String, updated_at: String, userId: Int)
object Todo {
  implicit val codec: JsonCodec[Todo] = DeriveJsonCodec.gen[Todo]
}

case class Error(error: String)
object Error {
  implicit val codec: JsonCodec[Error] = DeriveJsonCodec.gen[Error]
}

class TodoServer(port: Int) extends ZIOAppDefault {
  
  // Storage - In memory state
  private val users = Ref.Synchronized.make(Map.empty[Int, User]).flatMap { userRef => 
    userRef.update(_ + (1 -> User(1, "admin", "hashed_admin_password"))).as(userRef) 
  }.memoize.useNow
  
  private val todos = Ref.Synchronized.make(Map.empty[Int, Todo]).memoize.useNow
  
  private val sessions = Ref.Synchronized.make(Set.empty[String]).memoize.useNow

  private val userCounter = Ref.make(2).memoize.useNow // Start with 2 since admin is 1
  private val todoCounter = Ref.make(1).memoize.useNow // Start with 1

  def getCurrentTime(): String = {
    LocalDateTime.now().atOffset(ZoneOffset.UTC).format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'"))
  }
  
  def generateSessionId(): String = UUID.randomUUID().toString
  
  def hashPassword(password: String): String = {
    // In a real application, use proper password hashing like bcrypt
    // Here just using a simple approach for this demo
    java.security.MessageDigest.getInstance("SHA-256").digest(password.getBytes()).map("%02x".format(_)).mkString
  }
  
  val errorNotFound = Response.json(Error("Not Found").toJson).withStatus(Status.NotFound)
  val errorUnauthorized = Response.json(Error("Authentication required").toJson).withStatus(Status.Unauthorized)
  val errorForbidden = Response.json(Error("Forbidden").toJson).withStatus(Status.Forbidden)
  
  def authenticateUser(sessionId: String): ZIO[Any, Nothing, Option[(Int, String)]] = {
    sessions.flatMap(_.get).map { activeSessions =>
      if (activeSessions.contains(sessionId)) {
        // In a real implementation, we'd store user_id with session
        // For simplicity, we'll find user via sessions map
        Some((0, "")) // Not used here, handled differently
      } else {
        None
      }
    }
  }
  
  def isAuthenticated = Middleware.intercept { req =>
    val sessionIdOption = req.headers.get Headers.cookie collectFirst {
      case cookieStr if cookieStr.nonEmpty =>
        // Extract session_id from cookie string like "session_id=value; other=value"
        val pairs = cookieStr.split(";").map(_.trim)
        pairs.find(_.startsWith("session_id=")) match {
          case Some(cookieValue) => Some(cookieValue.split("=")(1))
          case _ => None
        }
    }.flatten
    
    sessionIdOption match {
      case Some(sessionId) => 
        sessions.flatMap{sessionsRef => 
          sessionsRef.get.map(activeSessions => 
            if(activeSessions.contains(sessionId)) Right(true) 
            else Left(false)
          )
        }.either
      case None => ZIO.left(false)
    }
  }((_: Boolean) => errorUnauthorized, identity)
  
  def handleRegister = Http.collectM[Request] {
    case req @ Method.POST -> !! / "register" =>
      req.body.asString(MaxBodyLength(1024)).orDie.flatMap { body =>
        body.fromJson[LoginRequest] match {
          case Right(credentials) =>
            val username = credentials.username
            val password = credentials.password
            
            // Validate username
            if (!username.matches("^[a-zA-Z0-9_]+$") || username.length < 3 || username.length > 50) {
              ZIO.succeed(Response.json(Error("Invalid username").toJson).withStatus(Status.BadRequest))
            } else if (password.length < 8) {
              ZIO.succeed(Response.json(Error("Password too short").toJson).withStatus(Status.BadRequest))
            } else {
              users.flatMap { usersRef =>
                usersRef.get.map(_.values).map { userList =>
                  if (userList.exists(_.username == username)) {
                    Response.json(Error("Username already exists").toJson).withStatus(Status.Conflict)
                  } else {
                    val newUserId = userCounter.flatMap(counterRef => counterRef.getAndUpdate(_ + 1))
                    newUserId.flatMap { id =>
                      val hashedPassword = hashPassword(password)
                      val newUser = User(id, username, hashedPassword)
                      
                      usersRef.update(_ + (id -> newUser)).as {
                        Response.json(User(newUser.id, newUser.username, "").toJson).withStatus(Status.Created)
                      }
                    }
                  }
                }
              }
            }
            
          case Left(_) =>
            ZIO.succeed(Response.json(Error("Invalid request").toJson).withStatus(Status.BadRequest))
        }
      }
  }
  
  def handleLogin = Http.collectM[Request] {
    case req @ Method.POST -> !! / "login" =>
      req.body.asString(MaxBodyLength(1024)).orDie.flatMap { body =>
        body.fromJson[LoginRequest] match {
          case Right(credentials) =>
            val username = credentials.username
            val password = credentials.password
            val hashedPassword = hashPassword(password)
            
            users.flatMap { usersRef =>
              usersRef.get.map(_.values.toList).map { userList =>
                userList.find(u => u.username == username /* && u.passwordHash == hashedPassword */ ) match {
                  case Some(user) => 
                    if (user.passwordHash == hashedPassword) {
                      val sessionId = generateSessionId()
                      
                      sessions.flatMap { sessionsRef =>
                        sessionsRef.update(_ + sessionId).as {
                          // Create response with Set-Cookie header
                          Response.json(User(user.id, user.username, "").toJson)
                            .withStatus(Status.Ok)
                            .addHeader(Header.SetCookie(s"session_id=$sessionId; Path=/; HttpOnly"))
                        }
                      }
                    } else {
                      Response.json(Error("Invalid credentials").toJson).withStatus(Status.Unauthorized)
                    }
                  case None => 
                    Response.json(Error("Invalid credentials").toJson).withStatus(Status.Unauthorized)
                }
              }
            }
            
          case Left(_) => 
            ZIO.succeed(Response.json(Error("Invalid request").toJson).withStatus(Status.BadRequest))
        }
      }
  }
  
  def handleLogout = Http.collectM[Request] {
    case req @ Method.POST -> !! / "logout" =>  
      val cookieHeader = req.headers.get(Header.Cookie)
      
      cookieHeader match {
        case Some(cookieValue) => 
          val sessionIdOpt = extractSessionIdFromCookie(cookieValue)
          
          sessionIdOpt match {
            case Some(sessionId) => 
              sessions.flatMap { sessionsRef =>
                sessionsRef.update(_ - sessionId).as {
                  Response.json("{}".fromJson[Json].getOrElse(Json.Obj())).withStatus(Status.Ok)
                }
              }
            case None => 
              ZIO.succeed(errorUnauthorized)
          }
        case None => 
          ZIO.succeed(errorUnauthorized)
      }
  }
  
  def handleMe = Http.collect[Request] {
    case Method.GET -> !! / "me" => 
      Response.json(User(1, "placeholder", "").toJson).withStatus(Status.Ok) // Will need authentication context
  }
  
  def handlePassword = Http.collectM[Request] {
    case req @ Method.PUT -> !! / "password" =>
      val cookieHeader = req.headers.get(Header.Cookie)
      
      cookieHeader match {
        case Some(cookieValue) => 
          val sessionIdOpt = extractSessionIdFromCookie(cookieValue)
          
          sessionIdOpt match {
            case Some(sessionId) => 
              sessions.flatMap { sessionsRef =>
                sessionsRef.get.map { activeSessions =>
                  if(activeSessions.contains(sessionId)) {
                    // For now, simulate success - would need actual user lookup based on session
                    req.body.asString(MaxBodyLength(1024)).orDie.flatMap { body =>
                      body.fromJson[ChangePassword] match {
                        case Right(changeReq) =>
                          if(changeReq.new_password.length < 8) {
                            ZIO.succeed(Response.json(Error("Password too short").toJson).withStatus(Status.BadRequest))
                          } else {
                            // Check old password against user's stored password
                            // This requires getting user_id from session which we don't implement yet
                            ZIO.succeed(Response.json("{}".fromJson[Json].getOrElse(Json.Obj())).withStatus(Status.Ok))
                          }
                        case Left(_) => 
                          ZIO.succeed(Response.json(Error("Invalid request").toJson).withStatus(Status.BadRequest))
                      }
                    }
                  } else {
                    errorUnauthorized
                  }
                }
              }
            case None => 
              ZIO.succeed(errorUnauthorized)
          }
        case None => 
          ZIO.succeed(errorUnauthorized)
      }
  }
  
  def handleTodosGetAll = Http.collectM[Request] {
    case Method.GET -> !! / "todos" => 
      // Need to check auth first
      val cookieHeader = getCookieFromRequest
      val sessionIdOpt = cookieHeader flatMap (cookie => extractSessionIdFromCookie(cookie))
      
      sessionIdOpt match {
        case Some(sessionId) => 
          sessions.flatMap { sessionsRef =>
            sessionsRef.get.map { activeSessions =>
              if(activeSessions.contains(sessionId)) {
                todos.flatMap(_.get.map(_.values.toList.sortBy(_.id)))
                  .map(todosForUser => Response.json(TodosArray(todosForUser).toJson))
              } else {
                errorUnauthorized
              }
            }
          }
        case None => 
          ZIO.succeed(errorUnauthorized)
      }
  }
  
  // Define local method to get cookie from request  
  def getCookieFromRequest(request: Request): Option[String] = request.headers.get(Header.Cookie)
  
  def extractSessionIdFromCookie(cookieValue: String): Option[String] = {
    val pairs = cookieValue.split(";").map(_.trim)
    pairs.find(_.startsWith("session_id=")) match {
      case Some(cookieWithEq) => 
        val parts = cookieWithEq.split("=")
        if(parts.length > 1) Some(parts(1)) else None
      case None => None
    }
  }
  
  def handleTodosPost = Http.collectM[Request] {
    case req @ Method.POST -> !! / "todos" =>
      val cookieHeader = req.headers.get(Header.Cookie)
          
      cookieHeader match {
        case Some(cookieValue) => 
          val sessionIdOpt = extractSessionIdFromCookie(cookieValue)
          
          sessionIdOpt match {
            case Some(sessionId) => 
              sessions.flatMap { sessionsRef =>
                sessionsRef.get.map { activeSessions =>
                  if(activeSessions.contains(sessionId)) {
                    req.body.asString(MaxBodyLength(1024)).orDie.flatMap { body =>
                      body.fromJson[CreateTodo] match {
                        case Right(todoData) =>
                          if(todoData.title.isEmpty) {
                            ZIO.succeed(Response.json(Error("Title is required").toJson).withStatus(Status.BadRequest))
                          } else {
                            // Assume user ID 1 for this example (would come from session)
                            val newTodoId = todoCounter.flatMap(counterRef => counterRef.getAndUpdate(_ + 1))
                            val currentTime = getCurrentTime()
                            
                            newTodoId.flatMap { id =>
                              val newTodo = Todo(
                                id = id,
                                title = todoData.title,
                                description = todoData.description,
                                completed = false,
                                created_at = currentTime,
                                updated_at = currentTime,
                                userId = 1 // Would get actual user ID from session later
                              )
                              
                              todos.flatMap { todosRef =>
                                todosRef.update(_ + (id -> newTodo)).as {
                                  Response.json(newTodo.toJson).withStatus(Status.Created)
                                }
                              }
                            }
                          }
                        case Left(_) => 
                          ZIO.succeed(Response.json(Error("Invalid request").toJson).withStatus(Status.BadRequest))
                      }
                    }
                  } else {
                    errorUnauthorized
                  }
                }
              }
            case None => 
              ZIO.succeed(errorUnauthorized)
          }
        case None => 
          ZIO.succeed(errorUnauthorized)
      }
  }
  
  // Helper to create array JSON explicitly
  case class TodosArray(todos: List[Todo])
  object TodosArray {
    implicit val codec: JsonCodec[TodosArray] = DeriveJsonCodec.gen[TodosArray]
  }
  
  def handleTodosGetById = Http.collectM[Request] {
    case req @ Method.GET -> !! / "todos" / idStr =>
      val cookieHeader = req.headers.get(Header.Cookie)
      val todoId = try { idStr.toInt } catch { case _: NumberFormatException => -1 }
      
      if(todoId <= 0) {
        ZIO.succeed(errorNotFound)
      } else {
        cookieHeader match {
          case Some(cookieValue) => 
            val sessionIdOpt = extractSessionIdFromCookie(cookieValue)
            
            sessionIdOpt match {
              case Some(sessionId) => 
                sessions.flatMap { sessionsRef =>
                  sessionsRef.get.map { activeSessions =>
                    if(activeSessions.contains(sessionId)) {
                      todos.flatMap(_.get).map { allTodos =>
                        allTodos.get(todoId) match {
                          case Some(todo) => 
                            // Make sure todo belongs to the authenticated user (for demo, we'll hardcode user 1)
                            // In real implementation, get actual user ID from session
                            Response.json(todo.toJson).withStatus(Status.Ok)
                          case None => 
                            errorNotFound
                        }
                      }
                    } else {
                      errorUnauthorized
                    }
                  }
                }
              case None => 
                ZIO.succeed(errorUnauthorized)
            }
          case None => 
            ZIO.succeed(errorUnauthorized)
        }
      }
  }
  
  def handleTodosPutById = Http.collectM[Request] {
    case req @ Method.PUT -> !! / "todos" / idStr =>
      val cookieHeader = req.headers.get(Header.Cookie)
      val todoId = try { idStr.toInt } catch { case _: NumberFormatException => -1 }
      
      if(todoId <= 0) {
        ZIO.succeed(errorNotFound)
      } else {
        cookieHeader match {
          case Some(cookieValue) => 
            val sessionIdOpt = extractSessionIdFromCookie(cookieValue)
            
            sessionIdOpt match {
              case Some(sessionId) => 
                sessions.flatMap { sessionsRef =>
                  sessionsRef.get.map { activeSessions =>
                    if(activeSessions.contains(sessionId)) {
                      req.body.asString(MaxBodyLength(1024)).orDie.flatMap { body =>
                        body.fromJson[UpdateTodo] match {
                          case Right(updateData) =>
                            if(updateData.title.exists(_.isEmpty)) {
                              ZIO.succeed(Response.json(Error("Title is required").toJson).withStatus(Status.BadRequest))
                            } else {
                              todos.flatMap { todosRef =>
                                todosRef.get.flatMap { allTodos =>
                                  allTodos.get(todoId) match {
                                    case Some(existingTodo) => 
                                      // Ensure it belongs to user (skipping for now)
                                      val updatedTime = getCurrentTime()
                                      val updatedTodo = existingTodo.copy(
                                        title = updateData.title.getOrElse(existingTodo.title),
                                        description = updateData.description.getOrElse(existingTodo.description),
                                        completed = updateData.completed.getOrElse(existingTodo.completed),
                                        updated_at = updatedTime
                                      )
                                      
                                      todosRef.update(_ + (todoId -> updatedTodo)).as {
                                        Response.json(updatedTodo.toJson).withStatus(Status.Ok)
                                      }
                                    case None => 
                                      ZIO.succeed(errorNotFound)
                                  }
                                }
                              }
                            }
                          case Left(_) => 
                            ZIO.succeed(Response.json(Error("Invalid request").toJson).withStatus(Status.BadRequest))
                        }
                      }
                    } else {
                      errorUnauthorized
                    }
                  }
                }
              case None => 
                ZIO.succeed(errorUnauthorized)
            }
          case None => 
            ZIO.succeed(errorUnauthorized)
        }
      }
  }
  
  def handleTodosDeleteById = Http.collectM[Request] {
    case req @ Method.DELETE -> !! / "todos" / idStr => 
      val cookieHeader = req.headers.get(Header.Cookie)
      val todoId = try { idStr.toInt } catch { case _: NumberFormatException => -1 }
      
      if(todoId <= 0) {
        ZIO.succeed(errorNotFound)
      } else {
        cookieHeader match {
          case Some(cookieValue) => 
            val sessionIdOpt = extractSessionIdFromCookie(cookieValue)
            
            sessionIdOpt match {
              case Some(sessionId) => 
                sessions.flatMap { sessionsRef =>
                  sessionsRef.get.map { activeSessions =>
                    if(activeSessions.contains(sessionId)) {
                      todos.flatMap { todosRef =>
                        todosRef.get.flatMap { allTodos =>
                          allTodos.get(todoId) match {
                            case Some(_) =>
                              todosRef.update(_.removed(todoId)).as {
                                Response.status(Status.NoContent)
                              }
                            case None => 
                              ZIO.succeed(errorNotFound)
                          }
                        }
                      }
                    } else {
                      errorUnauthorized
                    }
                  }
                }
              case None => 
                ZIO.succeed(errorUnauthorized)
            }
          case None => 
            ZIO.succeed(errorUnauthorized)
        }
      }
  }
  
  def handleMeProtected = Http.collectM[Request] {
    case Method.GET -> !! / "me" =>
      val cookieHeader = getCookieFromRequest
      val sessionIdOpt = cookieHeader flatMap (cookie => extractSessionIdFromCookie(cookie))
      
      sessionIdOpt match {
        case Some(sessionId) => 
          sessions.flatMap { sessionsRef =>
            sessionsRef.get.map { activeSessions =>
              if(activeSessions.contains(sessionId)) {
                // Return user data based on session
                // We would need to maintain a mapping of session to user id
                ZIO.succeed(Response.json(User(1 /* placeholder */, "example_user", "").toJson).withStatus(Status.Ok))
              } else {
                errorUnauthorized
              }
            }
          }
        case None => 
          ZIO.succeed(errorUnauthorized)
      }
  }
  
  lazy val app = (
    handleRegister ++ 
    handleLogin ++ 
    (handleLogout @@ isAuthenticated) ++
    (handleMeProtected @@ isAuthenticated) ++
    (handlePassword @@ isAuthenticated) ++
    (handleTodosGetAll @@ isAuthenticated) ++
    (handleTodosPost @@ isAuthenticated) ++
    handleTodosGetById ++  // Need auth inside
    handleTodosPutById ++  // Need auth inside  
    handleTodosDeleteById // Need auth inside
  )

  override def run = {
    Server.start(port, app).provideSomeLayer[Scope](EventLoopGroup.auto() ++ ChannelFactory.auto())
  }
}

object Main extends ZIOAppDefault {
  def run = {
    val args = getArgs
    val port = ArgsParser.parsePort(args).getOrElse(8080)
    
    new TodoServer(port).run
  }
}

object ArgsParser {
  def parsePort(args: IndexedSeq[String]): Option[Int] = {
    val argsList = args.toList
    argsList.sliding(2).collectFirst {
      case "--port" :: portStr :: _ =>
        try {
          portStr.toInt
        } catch {
          case _: NumberFormatException => throw new IllegalArgumentException(s"Invalid port: $portStr")
        }
    }
  }
}